### PR TITLE
feat: implement v4-quoter with computeV4Routes and stub getQuotes (invoke onchain quoter provider class)

### PR DESCRIPTION
**Pull request by jsy1218**
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature

- **What is the current behavior?** (You can also link to an open issue here)
After getting the top v4 candidate pools by TVL, we need to compute the v4-only routes.

- **What is the new behavior (if this is a feature change)?**
We are implementing computeV4Routes as part of v4-quoter, that will get invoked by alpha router.

- **Other information**:
It's protected under explicitly protocol version specification.

### CODE REVIEW

**Comment by jsy1218** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Uniswap/smart-order-router/666?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#666** <a href="https://app.graphite.dev/github/pr/Uniswap/smart-order-router/666?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#664** <a href="https://app.graphite.dev/github/pr/Uniswap/smart-order-router/664?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#661** <a href="https://app.graphite.dev/github/pr/Uniswap/smart-order-router/661?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @jsy1218 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/routers/alpha-router/alpha-router.ts [hunk 1]
```diff
@@ -1737,6 +1752,43 @@ export class AlphaRouter
...
+        this.v4Quoter
+          .getQuotes(
+            v4RoutesFromCache,
+            amounts,
+            percents,
+            quoteToken,
+            tradeType,
+            routingConfig,
+            undefined,
+            v3GasModel,
```
Comment (xrsv) [inline src/routers/alpha-router/alpha-router.ts:1777]: is this the same as v3 or will it change?

File: src/routers/alpha-router/quoters/base-quoter.ts [hunk 2]
```diff
@@ -41,10 +43,10 @@ import { GetQuotesResult, GetRoutesResult } from './model/results';
  */
 export abstract class BaseQuoter<
   CandidatePools extends
-    | V2CandidatePools
-    | V3CandidatePools
+    | SupportedCandidatePools
     | [V3CandidatePools, V2CandidatePools, CrossLiquidityCandidatePools],
```
Comment (xrsv) [inline src/routers/alpha-router/quoters/base-quoter.ts:47]: what about this line here? does it need updating with V4?

File: src/routers/alpha-router/quoters/base-quoter.ts [hunk 3]
```diff
@@ -200,14 +202,16 @@ export abstract class BaseQuoter<
     const tokens = _.flatMap(pools, (pool) => [pool.token0, pool.token1]);
 
     const tokenValidationResults =
-      await this.tokenValidatorProvider.validateTokens(tokens);
+      await this.tokenValidatorProvider.validateTokens(
+        tokens.map((token) => token.wrapped)
```
Comment (xrsv) [inline src/routers/alpha-router/quoters/base-quoter.ts:206]: curious why is this change needed here? will it affect V2 Pair?

File: src/routers/alpha-router/quoters/v4-quoter.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    _amounts: CurrencyAmount[],
+    _percents: number[],
+    _quoteToken: Token,
+    _tradeType: TradeType,
+    _routingConfig: AlphaRouterConfig,
+    _candidatePools: CandidatePoolsBySelectionCriteria | undefined,
+    _gasModel: IGasModel<RouteWithValidQuote> | undefined,
+    _gasPriceWei: BigNumber | undefined
+  ): Promise<GetQuotesResult> {
+    throw new Error('Method not implemented.');
```
Comment (xrsv) [inline src/routers/alpha-router/quoters/v4-quoter.ts:129]: i assume this will be implemented later?

**Review by xrsv — APPROVED** [review]
looks good! left some comments

File: src/routers/alpha-router/quoters/base-quoter.ts (hunk 2 above)
Comment (jsy1218) [reply src/routers/alpha-router/quoters/base-quoter.ts:47]: We don't want to update this line yet, because it's touching the mixed routing/quoting.

File: src/routers/alpha-router/quoters/base-quoter.ts (hunk 3 above)
Comment (jsy1218) [reply src/routers/alpha-router/quoters/base-quoter.ts:206]: Aw, this is because in V4, we will support native currency routing/quoting. In V2 and V3, all uniswap pools can only support ERC20 tokens, but not native currencies. In V4, any Uniswap pool can support native currency. So that means, data structure wise, tokens will change its data type from `Token` to `Currency`, where `Currency = Token | NativeCurrency`. This means, since V2 and V3 don't support native currency pool, so far we can just safely wrap the token with type `Currency` to type `Token`. But in v4, we will need to dedicate our effort in supporting native currency routing in separate PRs, since I expect that to be non-trivial amount of work.

File: src/routers/alpha-router/quoters/v4-quoter.ts (hunk 4 above)
Comment (jsy1218) [reply src/routers/alpha-router/quoters/v4-quoter.ts:129]: that's right. intentionally not implemented in this PR, so that it can be implemented in a dependent PR https://github.com/Uniswap/smart-order-router/pull/668

File: src/routers/alpha-router/alpha-router.ts (hunk 1 above)
Comment (jsy1218) [reply src/routers/alpha-router/alpha-router.ts:1777]: It will need to change. I have a https://linear.app/uniswap/issue/ROUTE-223/support-v4-gas-model ticket.

**Comment by jsy1218** [comment]
### Merge activity

* **Aug 21, 11:32 PM EDT**: @jsy1218 started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Uniswap/smart-order-router/666).
